Fix virtual static call

DragonSA opened the pull request:
## Description

Fix calling an overridden virtual static method via the interface.

The current fsharp calls, ends up calling the default implementation of C# static virtual methods.  This fixes this by ensuring that the static virtual metadata is honoured in method resolution.

## Checklist

- [X] Test cases added
- [x] Release notes entry updated

## Sponsorship
[Counterpoint Dynamics (Pty) Ltd](https://cpdynamics.co.za)

vzarytovskii commented:
What was the behaviour before? This needs SRTP tests as well. Also need to make sure RFC for IWSAMs mentions virtual members as well.

KevinRansom commented:
@vzarytovskii ... 

It compiles the method invocation however it invokes the base class method implementation and not the overridden version.

static virtual members are confusingly necessary for the INumerics stuff because of generics.

It's a good change.

vzarytovskii commented:
> @vzarytovskii ... 
> 
> 
> 
> It compiles the method invocation however it invokes the base class method implementation and not the overridden version.
> 
> 
> 
> static virtual members are confusingly necessary for the INumerics stuff because of generics.
> 
> 
> 
> It's a good change.
> 
> 

It's a breaking change then? What do we do in case of non-statics?

KevinRansom commented on tests/FSharp.Compiler.ComponentTests/Interop/StaticsInInterfaces.fs:
```diff
@@ -606,6 +606,36 @@ module Test =
+
+let echo<'T when 'T :> Test.I> x = 'T.Echo(x)
+
+if echo<Imp> "a" <> "a_imp" then
```
On order to make what this does a tiny bit clearer, I would modify this test case slightly:
````

type Imp() =
    interface Test.I with
        static member Echo (x: string) = x + "_imp"

let echo<'T when 'T :> Test.I> x = 'T.Echo(x)
let result = echo<Imp> "a"

if result = "a_imp" then
    printf $"Succeeded"
elif result = "a" then
    failwith "Incorrectly invoked the base class 'Echo'"
else
    failwith "incorrect value"
````

DragonSA replied:
Updated in b16cf8351

KevinRansom commented on tests/FSharp.Compiler.ComponentTests/Language/InterfaceTests.fs:
```diff
@@ -4,8 +4,8 @@ open Xunit
 [<Fact>]
-let ``Concrete instance method is not allowed in interfaces in lang preview``() =
-    FSharp $"""
+let ``Concrete instance method is not allowed in interfaces in lang version80``() =
```
Thank you for taking care of this, we do forget to do that from time to time.

KevinRansom replied:
@vzarytovskii, the current behavior would be unexpected, but yes, it would change the behavior of existing applications if they relied on the bug.  I think we would take this as a breaking change.  It's not actually clear to me that the INumerics would work as expected without this fix.

For example:
![image](https://github.com/dotnet/fsharp/assets/5175830/cdc6bf59-5893-4ebc-ad71-dc61e5547883)

Someone should probably take a look, perhaps I should.

DragonSA replied:
I was trying to implement `INumber.Max`, and came across this bug.  I was not able to get F# to call my defined `Max`, instead the base method keeps getting called.  The unit tests from [here](https://gitlab.com/dbnZA/Jita/-/merge_requests/67) pass with this change, but not without.

KevinRansom reviewed (commented):
The change is a good one, we need to consider the breaking change.  Although my intuition is that it is rare enough and breaks an important dotnet feature so we would take the break.

Although ... the scenario to implement the INumerics, may not have been a goal, I wasn't involved in those design discussions.

KevinRansom commented:
@vzarytovskii interfaces with virtual statics was implemented as part of the INumerics feature.  We handle abstract/virtuals/override on instance methods compatibly with C#.  I think we probably consume, the framework INumber<T> implementations correctly.  This particular bug is likely merely an oversight.

KevinRansom commented:
Oh yeah, and there is no real performance implications.

vzarytovskii commented:
> @vzarytovskii interfaces with virtual statics was implemented as part of the INumerics feature.  We handle abstract/virtuals/override on instance methods compatibly with C#.  I think we probably consume, the framework INumber<T> implementations correctly.  This particular bug is likely merely an oversight.

I know, I implemented those in F#.

Question is which method do we call now
A. In case of static virtual 
B. In case of non-static virtual

A should match B.

KevinRansom commented:
For virtual inheritence, it should be the most derived.  Static should emulate that I should think.

DragonSA commented:
> What was the behaviour before? This needs SRTP tests as well. Also need to make sure RFC for IWSAMs mentions virtual members as well.

The prior behaviour was the override of a virtual static method could not be called by F#.  

Just a note that SRTP cannot be used to call the override of `Echo`, since that is defined on the interface and not the class itself.  I've added a test (in 93e382869180f) to confirm the correct static methods get called.  

Regarding the RFC, are you referring to [RFC FS-1124](https://github.com/fsharp/fslang-design/blob/main/FSharp-7.0/FS-1124-interfaces-with-static-abstract-members.md), this does not mention virtual static methods.  

> > @vzarytovskii ...
> > It compiles the method invocation however it invokes the base class method implementation and not the overridden version.
> > static virtual members are confusingly necessary for the INumerics stuff because of generics.
> > It's a good change.
> 
> It's a breaking change then? What do we do in case of non-statics?

The overridden method is called:
```fsharp
    [<FactForNETCOREAPP>]
    let ``F# can call overwritten virtual member from interface``() =
        let CSharpLib =
            CSharp """
namespace Test;

public interface I
{
    virtual string Echo(string x) => $"I.Echo: {x}";
}
            """
            |> withCSharpLanguageVersion CSharpLanguageVersion.CSharp11
            |> withName "CsLibAssembly"

        FSharp """
type Imp() =
    interface Test.I with
        member _.Echo (x: string) = $"Imp.I.Echo: {x}"

let echo<'T when 'T :> Test.I> (a: 'T) x = a.Echo(x)

match echo (Imp()) "a" with
| "Imp.I.Echo: a" -> printfn "success"
| _ -> failwith "incorrect value"
"""
        |> withReferences [CSharpLib]
        |> withLangVersion80
        |> asExe
        |> compileAndRun
        |> shouldSucceed
```


> > @vzarytovskii interfaces with virtual statics was implemented as part of the INumerics feature.  We handle abstract/virtuals/override on instance methods compatibly with C#.  I think we probably consume, the framework INumber implementations correctly.  This particular bug is likely merely an oversight.
> 
> I know, I implemented those in F#.
> 
> Question is which method do we call now A. In case of static virtual B. In case of non-static virtual
> 
> A should match B.

I agree :-).  Currently we:
A. In case of static virtual: F# calls the base implementation (C# calls the override) (this change calls the override).  
B. In case of non-static: F# calls the override (C# does the same).  

Thus, this change is required to bring A and B into alignment.

DragonSA commented:
Out of interest, what is the protocol for resolving comments?  Should I do it once I made the appropriate change/comment, or does the commenter resolve it once satisfied?

vzarytovskii commented:
> Regarding the RFC, are you referring to RFC FS-1124, this does not mention virtual static methods.

Yeah, this one. Probably then add a one-liner saying what will be preferred in case of virtual methods.

DragonSA commented:
> > Regarding the RFC, are you referring to RFC FS-1124, this does not mention virtual static methods.
> 
> Yeah, this one. Probably then add a one-liner saying what will be preferred in case of virtual methods.

I've created fsharp/fslang-design#766

vzarytovskii commented:
> Out of interest, what is the protocol for resolving comments?  Should I do it once I made the appropriate change/comment, or does the commenter resolve it once satisfied?

Feel free to resolve if you've addressed it

edgarfgp commented:
Thanks @ DragonSA  and to `Counterpoint Dynamics (Pty)` for sponsoring this. Really cool 👍

vzarytovskii reviewed (approved):
Looks good, thanks

dawedawe commented:
> Thanks @DragonSA and to `Counterpoint Dynamics (Pty)` for sponsoring this. Really cool 👍

Absolutely, thanks for sponsoring :heart: 
Would be awesome to see more actions like this.
